### PR TITLE
[SPARK-21281][SQL] Use string types by default if array and map have no argument

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr modified code to use string types by default if `array` and `map` in functions have no argument. This behaviour is the same with Hive one;
```
hive> CREATE TEMPORARY TABLE t1 AS SELECT map();
hive> DESCRIBE t1;
_c0   map<string,string>                          

hive> CREATE TEMPORARY TABLE t2 AS SELECT array();
hive> DESCRIBE t2;
_c0   array<string>
```

## How was this patch tested?
Added tests in `DataFrameFunctionsSuite`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 1]
```diff
@@ -448,6 +448,43 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
...
+    }.getMessage
+    assert(errMsg.contains("due to data type mismatch: input to function coalesce cannot be empty"))
+
+    // This hits java.lang.AssertionError
+    // spark.range(1).select(struct())
+
+    errMsg = intercept[IllegalArgumentException] {
+      spark.range(1).select(greatest())
+    }.getMessage
+    assert(errMsg.contains("requirement failed: greatest requires at least 2 arguments"))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Nit: Exception types are different between `sql.functions.greatest` and `selectExpr("greatest()")`;
```
cala> spark.range(1).select(greatest()).show
java.lang.IllegalArgumentException: requirement failed: greatest requires at least 2 arguments.
  at scala.Predef$.require(Predef.scala:224)
  at org.apache.spark.sql.functions$.greatest(functions.scala:1570)
  ... 48 elided

scala> spark.range(1).selectExpr("greatest()").show
org.apache.spark.sql.AnalysisException: cannot resolve 'greatest()' due to data type mismatch: GREATEST requires at least 2 arguments; line 1 pos 0;
'Project [unresolvedalias(greatest(), Some(<function1>))]
+- Range (0, 1, step=1, splits=Some(4))

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:42)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$2.applyOrElse(CheckAnalysis.scala:95)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$2.applyOrElse(CheckAnalysis.scala:87)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 2]
```diff
@@ -448,6 +448,43 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       rand(Random.nextLong()), randn(Random.nextLong())
     ).foreach(assertValuesDoNotChangeAfterCoalesceOrUnion(_))
   }
+
+  test("SPARK-21281 fails if functions have no argument") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: We currently have the six functions that have variable-length arguments and cannot stand no argument in `sql.functions`.

**Comment by maropu** [comment]
The hive-related tests failed, so I checked the hive behaviour;
```
hive> CREATE TEMPORARY TABLE t1 AS SELECT map();
hive> DESCRIBE t1;
_c0   map<string,string>                          

hive> CREATE TEMPORARY TABLE t2 AS SELECT array();
hive> DESCRIBE t2;
_c0   array<string>

hive> CREATE TEMPORARY TABLE t3 AS SELECT coalesce();
FAILED: SemanticException [Error 10305]: CREATE-TABLE-AS-SELECT creates a VOID type, please use CAST to specify the type, near field:  _c0

hive> CREATE TEMPORARY TABLE t4 AS SELECT struct();
hive> DESCRIBE t4;
_c0                     struct<>   

hive> CREATE TEMPORARY TABLE t5 AS SELECT greatest();
FAILED: SemanticException [Error 10015]: Line 1:37 Arguments length mismatch 'greatest': greatest requires at least 2 arguments, got 0

hive> CREATE TEMPORARY TABLE t6 AS SELECT least();
FAILED: SemanticException [Error 10015]: Line 1:37 Arguments length mismatch 'least': least requires at least 2 arguments, got 0
```
In hive, `array` and `map` with no argument use string types by default. Spark fails when using `struct` with no argument though, it seems hive return an empty struct. If you get time, could you check this? @gatorsmile Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Could you move these functions to `.sql`?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: uh. I see.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 3]
```diff
@@ -41,8 +41,13 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
 
   override def foldable: Boolean = children.forall(_.foldable)
 
-  override def checkInputDataTypes(): TypeCheckResult =
-    TypeUtils.checkForSameTypeInputExpr(children.map(_.dataType), "function array")
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children == Nil) {
+      TypeCheckResult.TypeCheckFailure("input to function coalesce cannot be empty")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: `coalesce `?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 4]
```diff
@@ -168,7 +173,9 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
   override def foldable: Boolean = children.forall(_.foldable)
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (children.size % 2 != 0) {
+    if (children == Nil) {
+      TypeCheckResult.TypeCheckFailure("input to function coalesce cannot be empty")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: `coalesce ` ?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Could you create a helper function for removing these duplicate codes?

Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 4 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: oh, my bad.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 5]
```diff
@@ -247,11 +247,8 @@ abstract class HashExpression[E] extends Expression {
   override def nullable: Boolean = false
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (children.isEmpty) {
-      TypeCheckResult.TypeCheckFailure("function hash requires at least one argument")
-    } else {
-      TypeCheckResult.TypeCheckSuccess
-    }
+    TypeUtils.checkTypeInputDimension(
+      children.map(_.dataType), s"function $prettyName", requiredMinDimension = 1)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: Oh, now we have a consistent message instsead of `at least one argument` or `cannot be empty`. :)

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: ok, thanks

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 6]
```diff
@@ -292,18 +296,23 @@ trait CreateNamedStructLike extends Expression {
...
+        s"input to function $prettyName requires at least one argument")
     } else {
-      val invalidNames = nameExprs.filterNot(e => e.foldable && e.dataType == StringType)
-      if (invalidNames.nonEmpty) {
-        TypeCheckResult.TypeCheckFailure(
-          "Only foldable StringType expressions are allowed to appear at odd position, got:" +
-          s" ${invalidNames.mkString(",")}")
-      } else if (!names.contains(null)) {
-        TypeCheckResult.TypeCheckSuccess
+      if (children.size % 2 != 0) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: We can use `else if` to flatten the nest `if`.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: ok

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by gatorsmile — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 7]
```diff
@@ -292,14 +296,17 @@ trait CreateNamedStructLike extends Expression {
   }
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (children.size % 2 != 0) {
+    if (children.length < 1) {
+      TypeCheckResult.TypeCheckFailure(
+        s"input to function $prettyName requires at least one argument")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: This is not related to what this PR claims to do. What's the reason behind this change?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: This is a behavior change and caused a problem in https://github.com/apache/spark/pull/22373

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala]: sorry, but I don't remember correctly.
I looked over this pr again and I also think the modification is not related to this pr.
